hydroshare_jupyter_sync/lib/filesystem/fs_map.py

```diff
@@ -0,0 +1,217 @@
+from abc import ABC, abstractmethod
+from collections import UserDict
+from pathlib import Path
+from typing import Dict, List, Union
+from hsclient import HydroShare
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+from .fs_resource_map import (
+    RemoteFSResourceMap,
+    FSResourceMap,
+    LocalFSResourceMap,
+)
+
+from .types import ResourceId
+
+
+# Abstract Interfaces
+
+
+class IFSMap(ABC):
+    """Abstract interface describing a FSMap's verbs (methods)"""
+
+    @classmethod
+    @abstractmethod
+    def create_map(cls) -> "FSMap":
+        """Create a new file system map instance from a provided root location.
+        Note, subclasses overload this method."""
+
+    @abstractmethod
+    def add_resource(self, resource_id: ResourceId) -> None:
+        """Create new FSResourceMap and add to the FSMap instance"""
+
+    @abstractmethod
+    def delete_resource(self, resource_id: ResourceId) -> None:
+        """Remove FSResourceMap instance from FSMap instance"""
+
+    @abstractmethod
+    def update_resource(self, resource_id: ResourceId) -> None:
+        """Update FSResourceMap instance in FSMap instance"""
+
+
+class IEntityFSMap(ABC):
+    """Abstract interface describing FSMap verbs (methods) that operate on FS entities."""
+
+    @abstractmethod
+    def add_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        """Remove a file in a FSResourceMap for a given resource id."""
+
+    @abstractmethod
+    def delete_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        """Remove a file in a FSResourceMap for a given resource id."""
+
+    @abstractmethod
+    def update_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        """Update a file in a FSResourceMap for a given resource id."""
+
+
+# Semi-concrete implementations
+
+
+class FSMap(UserDict, IFSMap, ABC):
+    """Class representing the relationship between HydroShare resource's, resource files, and
+    resource MD5 Hashes."""
+
+    def delete_resource(self, resource_id: ResourceId) -> None:
+        """Remove FSResourceMap instance from FSMap instance"""
+        if resource_id in self.data:
+            del self.data[resource_id]
+
+    def update_resource(self, resource_id: ResourceId) -> None:
+        """Update the FSResourceMap for a given resource id."""
+        if resource_id in self.data:
```

Comment:
On all these methods, especially for update/add, you'll want to know if the resource isn't in the map.

Reply:
including the localfsmap below

Reply:
Right, I went back and forth on this. I am happy to add a custom key/value error, I was just writing to follow the norms of the `dict` class in python.

Reply:
a custom error sounds good to me